### PR TITLE
Resource to manage machine learning jobs

### PR DESCRIPTION
**Pull request by csmarchbanks**
Draft PR to create a terraform resource for creating, updating, and deleting machine learning jobs. The client has not been pushed quite yet while it undergoes a bit of review, but figured we could start reviewing this in the meantime.

### CODE REVIEW

File: grafana/resource_machine_learning_job.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"datasource_id": {
+				Description: "The id of the datasource to query.",
+				Type:        schema.TypeInt,
+				Required:    true,
+			},
+			"datasource_type": {
+				Description: "The type of datasource to use when querying. Currently allowed values are prometheus, graphite, loki, postgres, and datadog.",
```
Comment (sd2k) [inline grafana/resource_machine_learning_job.go]:
```suggestion
				Description: "The type of datasource being queried. Currently allowed values are prometheus, graphite, loki, postgres, and datadog.",
```

File: grafana/resource_machine_learning_job.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+				Required:    true,
+			},
+			"interval": {
+				Description: "The data interval in seconds to train the data on.",
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     300,
+			},
+			"hyper_params": {
+				Description: "The hyper parameters used to fine tune the algorithm. See https://grafana.com/docs/grafana-cloud/machine-learning/models/ for the full list of available hyper parameters.",
```
Comment (sd2k) [inline grafana/resource_machine_learning_job.go]: For consistency:

```suggestion
				Description: "The hyperparameters used to fine tune the algorithm. See https://grafana.com/docs/grafana-cloud/machine-learning/models/ for the full list of available hyperparameters.",
```

File: grafana/resource_machine_learning_job_test.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+					resource.TestCheckResourceAttr("grafana_machine_learning_job.test_job", "query_params.expr", "grafanacloud_grafana_instance_active_user_count"),
+					resource.TestCheckResourceAttr("grafana_machine_learning_job.test_job", "interval", "300"),
+					resource.TestCheckResourceAttr("grafana_machine_learning_job.test_job", "training_window", "7776000"),
+					resource.TestCheckResourceAttr("grafana_machine_learning_job.test_job", "hyper_params.daily_seasonality", "15"),
+					resource.TestCheckResourceAttr("grafana_machine_learning_job.test_job", "hyper_params.weekly_seasonality", "10"),
+				),
+			},
+		},
+	})
+}
```
Comment (sd2k) [inline grafana/resource_machine_learning_job_test.go:54]: If it's easy to do, it'd be great to have a failing test case here; something like an invalid datasource type or a job trying to train on too many samples, to show that error messages are useful.

**Review by sd2k — CHANGES_REQUESTED** [review]
Generally LGTM, just a few small suggestions and test request if it's easy to add.

File: grafana/resource_machine_learning_job_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccResourceMachineLearningJob(t *testing.T) {
+	CheckCloudTestsEnabled(t)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheckCloud(t) },
+		ProviderFactories: testAccProviderFactories,
```
Comment (julienduchesne) [inline grafana/resource_machine_learning_job_test.go:20]: For completeness, you can have: 
- a `CheckDestroy` test that checks that the job is actually destroyed after the test: See example -> `testAccFolderCheckDestroy`
- a custom check to test that the job is actually deployed (not just Terraform blindly setting some attributes): See example -> `testAccFolderCheckExists`

**Review by julienduchesne — APPROVED** [review]
LGTM! Very nice work! I have one comment but it's not a blocker. Up to you to decide if you want to implement these additional checks. Let me know!

File: grafana/resource_machine_learning_job_test.go (hunk 4 above)
Comment (csmarchbanks) [reply grafana/resource_machine_learning_job_test.go:20]: Awesome, thanks for pointing those checks out! I have added a second commit implementing both of them.

**Review by sd2k — APPROVED** [review]
LGTM! 👍

**Comment by csmarchbanks** [comment]
@julienduchesne All other dependencies have been merged, so this is ready to go now!

**Pull request merged**